certora/specs/erc4626.spec

```diff
@@ -0,0 +1,790 @@
+//     assert shares ==0 || assets == 0,"either shares or assets to be supplied";
+//     assert balanceCheck(toUnderlying, _ULBal, ULBal_, _ATBal, ATBal_, assetsReceived),"assets should be transferred correctly";
+// }
+
+
+/////////////////////////////// OTHER ERC4626 FUNCTIONS ////////////////////////////////////
+
+rule totalAssetCheck(){
+    totalAssets@withrevert();
+    assert !lastReverted;
+}
+
+// ghost to track sum of balances of atoken contract and check is equal to totalSupply
+// rule totalAssetCheck(){
+
+// }
+
+// should be able to deposit an amount as long as it's less than maxdeposit amount
+// rule maxDepositCheck(){
+//     maxDeposit@withrevert();
+//     assert !lastReverted;
+// }
+
+// 
+// rule maxdepLimit(){
+//     uint256 asset;
+//     deposit@withrevert(assets);
+    
+//     assert !lastReverted => assets<= maxDeposit();
+//     assert !lastReverted && assets == maxuint256 => maxDeposit() == maxuint;
+// }
+
+/***
+* rule to check the following for the covertToAssets function:
+* 1. Independent of the user
+* 2. No revert unless overflow
+* 3. Must round down
+*/
+// STATUS: WIP
+// rule convertToAssetsCheck(){
+//     env e1;
+//     env e2;
+//     env e3;
+//     uint256 shares1;
+//     uint256 shares2;
+//     uint256 assets1;
+//     uint256 assets2;
+//     uint256 assets3;
+//     uint256 combinedAssets;
+//     storage before  = lastStorage;
+    
+//     assets1         = convertToAssets@withrevert(e1, shares1)           at before;
+//     assets2         = convertToAssets@withrevert(e2, shares1)           at before;
+//     assets3         = convertToAssets@withrevert(e2, shares2)           at before;
+//     combinedAssets  = convertToAssets@withrevert(e3, shares1 +shares2)  at before;
+
+//     // assert !lastReverted,"should not revert except for overflow";
+//     assert assets1 == assets2,"conversion to assets should be independent of env such as msg.sender";
+//     assert assets1 + assets3 <= combinedAssets,"conversion should round down and not up";
+// }
+
+/***
+* rule to check the following for the covertToShares function:
+* 1. Independent of the user
+* 2. No revert unless overflow
+* 3. Must round down
+*/
+// STATUS: WIP
+rule convertToSharesCheck(){
+    env e1;
+    env e2;
+    env e3;
+    uint256 assets1;
+    uint256 assets2;
+    uint256 shares1;
+    uint256 shares2;
+    uint256 shares3;
+    uint256 combinedShares;
+    storage before  = lastStorage;
+    
+    shares1         = convertToShares@withrevert(e1, assets1)            at before;
+    shares2         = convertToShares           (e2, assets1)            at before;
+    shares3         = convertToShares           (e2, assets2)            at before;
+    combinedShares  = convertToShares           (e3, assets1 + assets2)  at before;
+
+    
+    assert shares1 == shares2,"conversion to shares should be independent of env such as msg.sender";
```

Comment:
The assertion should hold if all envs have the same index.

Reply:
True. The rule only checks that the return value is independent of the caller.